### PR TITLE
[inductor] speedup faketensorprop a bit

### PR DESCRIPTION
**Pull request by shunting314**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #112209

Horace mentioned that FakeTensorProp takes non-trivial time during torch.compile. I did some profiling and here is a cProfile dump for a BertForMaskedLM training compilation job : https://www.svgviewer.dev/s/N6NJcEut . The profile shows a large portion of FakeTensorProp is spent on snapshot_fake function.

I added some temporary dynamo_timed decorator and get the following information: 
Screenshot 2023-10-26 at 2 21 26 PM

- fake_tensor_prop takes 5.4s within which snapshot_fake takes 2s.
- Since snapshot_fake just does `t.detach()`. If we only call `t.detach()` when `t.requires_grad()` is true, we can reduce fake_tensor_prop to 3.5s and snapshot_fake to 0.15s.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @bdhirsh @ezyang since I'm not sure if it's safe to skip `t.detach()` call here.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

**Comment by Chillee** [comment]
I don't think this is actually safe - the main reason for the detach is to avoid situations like
```
a = ...
a.squeeze_()
```
inplace mutating the faketensor metadata.

**Review by ezyang — CHANGES_REQUESTED** [review]
per horace

**Comment by shunting314** [comment]
Thanks @Chillee / @ezyang , that makes sense.

As i understand, the detach call has the following 2 effects
1. detach the tensor from the computation graph
1. return a new tensor with a separate copy of metadata (storage is still shared with the original tensor).

Actually we only need the 2nd effect if the tensor does not requires grad. Do we already have some way to copy a tensor's metadata without going thru the dispatcher? If not should we create such an API and use it in `snapshot_fake` function.

**Comment by Chillee** [comment]
@shunting314 I don't think the first part is actually the expensive part in this case - I think it's just that it goes through fake tensor prop again.

**Comment by ezyang** [comment]
Detach actually, in principle, should be pretty fast. As in, we should be willing to have that entirely implemented in C++ for the most part. I'm not sure why it is so expensive.

**Comment by Chillee** [comment]
@ezyang I think it's since it goes through an entire fake tensor dispatch.
